.github/workflows/scripts/registry-login.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+#
+# Copyright 2023 SLSA Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+set -euo pipefail
+
+UNTRUSTED_IMAGE_OR_REPO="$1"
+UNTRUSTED_INPUT_USERNAME="$2"
+UNTRUSTED_SECRET_USERNAME="$3"
+UNTRUSTED_PASSWORD="$4"
+GCP_ACCESS_TOKEN="$5"
+
+DEFAULT_REGISTRY="docker.io"
+
+# Extract the registry from the $UNTRESTED_IMAGE_OR_REPO argument
+# See: https://stackoverflow.com/questions/37861791/how-are-docker-image-names-parsed#37867949
+#
+# Possible inputs and their corresponding registries:
+# 1. "gcr.io/my-project/my-image" -> "gcr.io"
+# 2. "localhost:5000/my-project/my-image" -> "localhost:5000"
+# 3. "my-project/my-image" -> "docker.io"
+# 4. "gcr.io" -> "gcr.io"
+if [[ "$UNTRUSTED_IMAGE_OR_REPO" == *"/"* ]]; then
+  # Extract possible domain (containing '.' or ':') as registry
+  untrusted_registry=$(echo "${UNTRUSTED_IMAGE_OR_REPO}" | cut -f1 -d "/" | { grep -E "\.|:" || true; })
+  if [ "${untrusted_registry}" == "" ]; then
+    # Use default registry if no domain is found
+    untrusted_registry="$DEFAULT_REGISTRY"
+  fi
+else
```

Comment:
IIUC, if your script can't parse the registry against the regex `*/*`, then fallback to the user-provided input.
This seems like different from the original behavior. I recommend re-using the original logic, so we can be more sure we're not regressing.

I know it's difficult to test workflows end-to-end on your own forks. The best we've had so far is to merge, and then wait for the e2e tests in slsa-framework/example-package to run. You might still be able to partially test this workflow on your own fork, however, after renaming a few references to point to your own fork.

Reply:
Thanks for the heads up. This approach is indeed safer.